### PR TITLE
 pkg/tlsf: Fix the way system functions are overriden.

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

_I screwed up_

The correct way to overrride the malloc family of functions in newlib-nano is to provide the *_r (reentrant) variants. Newlib implements the "normal" functions on top of these (see the newlib source code). Also, internally it calls the *_r functions when allocating buffers.

If only the "normal" non-reentrant functions are provided this will mean that some of the code will still use the vanilla newlib allocator. Furthermore, if one uses the whole heap as a pool for TLSF then the system may in the best case crash as there is no enough memory for its internall allocations or in the worst case function eratically (this depends on how the heap reserved, there is an upcomming series of commits in that direction).

This commit splits the handling between newlib and native. It also prepares the ground for future work on the pool initialization.

Right now I could only test this in ARM and native and I cannot ensure it will work on other platforms.

> _Replacing the system's memory allocator is not something that can be taken lightly and will inevitably require diving into the depths of the libc. Therefore I would say that using TLSF as a system wide allocator is ATM supported officially only on those plaftorms._

### Reviewing

This looks like lot of code but most of it is just renaming things an moving stuff around.

There is a bit of duplicate code but the alternative was to either do a hack with macros or litter the code with ifdefs, which can only get worse in my next PR where I add initialization routines that differ according to the platform.

### Testing procedure

Aside from reading the newlib sources, you can see the issue in a live system using the debugger.

Compile any example (with or without tlsf-malloc), grab a debugger and place a breakpoint in sbrk and _sbrk_r. Doing a backtrace will reveal it gets called by _malloc_r.

### Issues/PRs references

Original PR:  #9006

### CODE REVIEW

**Comment by jcarrano** [comment]
@cladmi Would you care to take a look at the build system part?

File: pkg/tlsf/Makefile.dep [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ifneq (,$(filter tlsf-malloc,$(USEMODULE)))
+  ifneq (,$(filter newlib,$(USEMODULE)))
+    USEMODULE += tlsf-malloc_newlib
+  else
+    USEMODULE += tlsf-malloc_native
```
Comment (cladmi) [inline pkg/tlsf/Makefile.dep]: Maybe here enforce that `BOARD` == `native`.
Would prevent being in this branch with other boards that do not have `newlib`.

Not sure what to do for other boards.

File: pkg/tlsf/contrib/Makefile.include [hunk 2]
```diff
@@ -0,0 +1 @@
+UNDEF += $(BINDIR)/tlsf-malloc/tlsf-malloc.o
```
Comment (cladmi) [inline pkg/tlsf/contrib/Makefile.include]: I think the `UNDEF` is not necessary for this step. I do not see any `weak` or putting in specific sections.

Comment (jcarrano) [reply pkg/tlsf/contrib/Makefile.include]: D'oh, that should have gone in a later commit.

**Comment by jcarrano** [comment]
I nobody can review TLSF then I propose we delete it. There are two issues, though:

- Removing the entire package would require reworking Lua.
- If we remove only tlsf-malloc I'm afraid someone will reintroduce it at some point in the future and make this same mistakes all over again.

**Comment by gschorcht** [comment]
@jcarrano I'm out of office. I will try to take a look on it tomorrow.

**Comment by benpicco** [comment]
So just to clarify: On newlib systems the reentrant versions are to be provided whereas on native (glibc) the non-reentrant  have to be implemented.

Then this PR is pretty straightforward. Maybe add a comment why it is not sufficient to also just implement the reentrant versions on native.

**Comment by benpicco** [comment]
I tested this both on `esp32-wroom-32` and `nucleo-l031k6` with `tests/malloc`, both ran fine. However, it appears as if the same binary is generated no matter if I set `USEPKG += tlsf` or `USEMODULE += tlsf-malloc` or neither.

I also ran `examples/lua_basic` and `examples/lua_REPL` on the esp32.
The basic example works fine, the REPL example loops at

```
 Welcome to the interactive interpreter
 L> 
 Exited. status: No errors, return code 0
 This is Lua: starting interactive session
 …
```

but it also does that on `master`.

**Comment by jcarrano** [comment]
@benpicco the lua examples are not affected by this, since they use tslf functions directly (i.e. they do not replace the system malloc).

The issue with `tests/malloc` is weird. I do not get the same results with `examples/default`. Let me investigate.

**Comment by jcarrano** [comment]
@benpicco I found the issue: [I deleted the UNDEF](https://github.com/RIOT-OS/RIOT/compare/d21cd79f586b0bc9fac9fe1acd630788fcd3b541..8e6a4ffe2318b988635960f609dc4926a59e0fc3#diff-8f916ddbda05416220232f958019ae11L1).

You can see nothing gets overriden by doing:

```
arm-none-eabi-nm -l bin/nucleo-l031k6/default.elf | grep _malloc_r
```

which is wrong.

I'm pushing the fix.

**Comment by benpicco** [comment]
Hm, I tested it again with `nucleo-l031k6` and `esp32-wroom-32`.
Now with `USEMODULE += tlsf-malloc` a different binary is generated for both platforms.

I remembered that without #12032 a manual initialization of the heap is needed, so I added
```diff
--- a/tests/malloc/main.c
+++ b/tests/malloc/main.c
@@ -70,8 +70,12 @@ void free_memory(struct node *head)
     }
 }
 
+#include "tlsf-malloc.h"
+static char heap[4096];
 int main(void)
 {
+    tlsf_add_global_pool(heap, sizeof(heap));
+
     while (1) {
         struct node *head = malloc(sizeof(struct node));
         total += sizeof(struct node);
```

But when I flash it, neither of the two boards will print anything.

File: pkg/tlsf/contrib/native.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#if defined(__GNUC__) && !defined(__clang__)    /* Clang supports __GNUC__ but
+                                                 * not the alloc_size()
+                                                 * attribute */
+
+#define ATTR_MALLOC  __attribute__((malloc, alloc_size(1)))
+#define ATTR_CALLOC  __attribute__((malloc, alloc_size(1,2)))
+#define ATTR_MALIGN  __attribute__((alloc_align(1), alloc_size(2), malloc))
+#define ATTR_REALLOC  __attribute__((alloc_size(2)))
+
+#else /* No GNU C -> no alias attribute */
```
[the remark this answers is not in the document]
Comment (jcarrano) [reply pkg/tlsf/contrib/native.c]: The comment is wrong, it should read:
```suggestion
#else /* No GNU C -> no *alloc attribute */
```

The warning is not necessary: the "alloc" attributes are just hints to the compiler. In fact, I'm not 100% sure they help here.

**Comment by jcarrano** [comment]
I tested with BOARD=nucleo-f103rb and it works for me.

**Comment by benpicco** [comment]
Ok, it's also working on `samr21-xpro` and `same54-xpro`.

On `esp8266-esp-12x` I get 

```
/opt/esp/newlib-xtensa/xtensa-lx106-elf/lib/libc.a(lib_a-mallocr.o): In function `_malloc_r':
/root/newlib/xtensa-lx106-elf/newlib/libc/stdlib/../../../.././newlib/libc/stdlib/mallocr.c:2328: multiple definition of `_malloc_r'
/home/benpicco/dev/RIOT/tests/malloc/bin/esp8266-esp-12x/tlsf-malloc/newlib.o:newlib.c:(.text._malloc_r+0x10): first defined here
collect2: error: ld returned 1 exit status
```

but I suppose this might be fixed by #11108

I guess `nucleo-l031k6` then just has too little RAM (although it fails in a strange way) and on esp comes with it's very own newlib and might do strange things there. @gschorcht may know more about it.

**Comment by benpicco** [comment]
Please squash.
The PR makes sense and is working.

Let's not block this on unrelated breakage.

**Comment by jcarrano** [comment]
@benpicco done

**Comment by mguetschow** [comment]
`USEMODULE=tlsf-malloc make -C examples/basic/hello-world BOARD=nrf52840dk all` fails on current `master` and on `2019.10-RC1` for me with

```
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-nano-mallocr.o): in function `_malloc_r':
/home/pere/src/newlib-salsa/build_nano/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/stdlib/../../../../../../../../newlib/libc/stdlib/nano-mallocr.c:235: multiple definition of `_malloc_r'; /home/mikolai/TUD/Code/RIOT-build/examples/basic/hello-world/bin/nrf52840dk/tlsf-malloc/newlib.o:/home/mikolai/TUD/Code/RIOT-build/pkg/tlsf/contrib/newlib.c:56: first defined here
collect2: error: ld returned 1 exit status
```

Not sure how this function overwrite should have worked before? I guess it may depend on the newlib / toolchain version?

**Pull request merged**